DOCS: New Timing and Latency section (DOCF-1179) (#1994)
Unity-Technologies/InputSystem: Packages/com.unity.inputsystem/Documentation~/TimingOptimizeForFixedUpdate.md

> # Optimize for fixed-timestep or physics-based scenarios

Reviewer: I wonder if this section could be divided into two pages somehow - there is a mixture of conceptual and task based information here, which makes it tricky to navigate. I'm thinking the stuff from line 47 could be the task content, and the stuff before would be the conceptual info.

I also wonder whether merging the info from the non-physics based stuff would be a good idea or not. You know the topic better but I wonder whether you could have an introduction to optimizing input events which explains both approaches, and then separate pages on the task content.

I'll leave it to you to decide though!

> If you are working with the physics system or using `FixedUpdate` to control your game in a scenario where a small amount of input latency is acceptable (for example, a few frames), the simplest approach is to set the input system update mode to **Process Events in Fixed Update**. This means your input code in `FixedUpdate` will operate as expected.

Reviewer: How do you set this setting? Can you link to the docs on this?

> However, if you are working with the physics system or using `FixedUpdate` to control your game and want to get the minimum possible latency from the Input System, minimizing lag, set the input system update mode to **Process Events in Dynamic Update**. However in doing this, you must understand how to avoid the problems which can arise when using this strategy. Although it might seem incorrect if you have code in `FixedUpdate`, for most cases, this approach minimizes lag compared with processing events in Fixed Update. The reasons for this are explained in detail on this page. Having a good understanding of how input is processed in each mode allows you to make your own decision about how best to process input for your particular project.

Reviewer: ```suggestion
To get the minimum possible latency from the Input System and minimize lag, set the input system update mode to **Process Events in Dynamic Update**. However in doing this, you must understand how to avoid the problems which can arise when using this strategy. Although it might seem incorrect if you have code in `FixedUpdate`, for most cases, this approach minimizes lag compared with processing events in Fixed Update. The reasons for this are explained in detail on this page. Having a good understanding of how input is processed in each mode allows you to make your own decision about how best to process input for your particular project.
```

Reviewer: TBH I'm not a massive fan of this section. How do you set the update mode? The last couple of sentences are a bit vague and not necessary too. I'd consider rewording this or reordering the info on the page.

## Input in Fixed Update mode

When you set the input system update mode to **Process Events in Fixed Update**, input events are processed in groups according to whether their timestamp falls within the current fixed time step. There might be none, one, or multiple `FixedUpdate` calls processed per frame depending on how fast the game's frame rate is updating compared with the fixed update time step period.

If your game’s frame rate is running faster than the fixed time step period, you will have either zero or one `FixedUpdate` call per frame, depending whether the previous fixed time step has completed or not. If your game’s frame rate is running slower than the fixed time step period, you will have one or more `FixedUpdate` calls per frame \- as many as are required to catch up with the number of completed fixed time step periods that have elapsed.

> You can learn more about how the player loop processes fixed update time steps in [Time And Frame Rate Management](https://docs.unity3d.com/Manual/TimeFrameManagement.html).

Reviewer: You can use an xref type link to link to pages in the user manual BTW https://docs.unity3d.com/Packages/com.unity.package-manager-doctools@3.6/manual/manual-xrefs.html

e.g `[Time And Frame Rate Management](xref:TimeFrameManagement).`

### When the frame rate runs faster than fixed time step duration

![image alt text](./Images/TimingFastFPS.png)

This diagram shows the frame rate running faster than the fixed update time step rate. Time progresses to the right, each frame is numbered, and shows its `Update` call at the start of the frame in orange. The fixed time step here is 0.02 seconds (50 times per second), and the game is running faster, at about 80 frames per second. In this situation there are some frames with one fixed update call, and some frames with none, depending on whether a full fixed update time step has completed by the time the frame starts. The fixed time step periods are marked with letters A, B, C, D, E, and the frames in which their corresponding fixed update calls occur are marked in green. The fixed update call for time step A occurs at the start of frame 4, the FixedUpdate call for time step B occurs at the start of frame 7, and so on.

> ### When the frame rate runs slower than the fixed time step duration:

Reviewer: ```suggestion
### When the frame rate runs slower than the fixed time step duration
```

This diagram shows the frame rate running faster than the fixed update time step rate. Each frame is numbered (1 to 7), and shows its `Update` call at the start of the frame in orange. The fixed time step here is 0.02s (50 steps per second), and the game is running faster, at about 80 frames per second. In this situation there are some frames with one fixed update call, and some frames with none.

The diagram shows an input event (shown as a blue dot) that occurs during some unprocessed time during frame 3. Although the event occurs during frame 3, it's not processed in the fixed update input processing step at the start of frame 4 because it didn't occur during fixed time step (**A**). Instead, it occurred during fixed time step (**B**), which is processed at the start of frame 7 \- the first frame to occur after fixed time step (**B**) has completely elapsed.

This has the counterintuitive effect that the processing of input on frame 4 actually ignores some input that has already occurred on frame 3, because it's only processing events that occurred in the last complete fixed time step: (**A**).

## Minimize latency when using fixed update code

> In order to minimize input latency in input code in your `FixedUpdate` calls, you should set the input system update mode to **Process Events in Dynamic Update**. This has the effect of eliminating the problem of unprocessed time described previously. You can then use an event-driven or polling technique to read your input without missing events that occurred after the last fixed timestep but before the current frame.

Reviewer: ```suggestion
To minimize input latency in input code in `FixedUpdate` calls, set the input system update mode to **Process Events in Dynamic Update**, which eliminates the problem of unprocessed time. You can then use an event-driven or polling technique to read input without missing events that occurred after the last fixed timestep but before the current frame.
```

> This can introduce the problem of missed or duplicate discrete events however (such as attempting to read whether a button was pressed in a given frame), so if you use this strategy, you must understand how to [avoid missed or duplicate events](TimingAvoidMissedOrDuplicateEvents.md) in mixed timing scenarios requiring fixed and dynamic input.

Reviewer: ```suggestion
However, the **Process Events in Dynamic Update** mode might introduce the problem of missed or duplicate discrete events, such as attempting to read whether a button was pressed in a given frame. If you use this strategy, you must understand how to [avoid missed or duplicate events](TimingAvoidMissedOrDuplicateEvents.md) in mixed timing scenarios requiring fixed and dynamic input.
```
